pkg/api/serve.go

```diff
@@ -71,25 +68,70 @@ func Serve(
 		actions,
 		logger,
 	)
-	controller.Configure(api)
+	HandlerFromMuxWithBaseURL(controller, apiRouter, BaseURL)
 
-	apiHandler := api.Serve(func(handler http.Handler) http.Handler {
-		// build handler for our REST API
-		return httputil.LoggingMiddleware(
-			RequestIDHeaderName,
-			logging.Fields{"service_name": LoggerServiceName},
-			promhttp.InstrumentHandlerCounter(requestCounter,
-				MetricsHandler(api.Context(),
-					NewCookieAPIHandler(handler))))
-	})
 	uiHandler := NewUIHandler(authService, gatewayDomain)
+	r.Mount("/_health", httputil.ServeHealth())
+	r.Mount("/metrics", promhttp.Handler())
+	r.Mount("/_pprof/", httputil.ServePPROF("/_pprof/"))
+	r.Mount("/", uiHandler)
+	return r
+}
+
+// OapiRequestValidatorWithOptions Creates middleware to validate request by swagger spec.
+// This middleware is good for net/http either since go-chi is 100% compatible with net/http.
+func OapiRequestValidatorWithOptions(swagger *openapi3.Swagger, options *openapi3filter.Options) func(http.Handler) http.Handler {
+	router, err := legacy.NewRouter(swagger)
+	if err != nil {
+		panic(err)
+	}
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// validate request
+			if statusCode, err := validateRequest(r, router, options); err != nil {
+				http.Error(w, err.Error(), statusCode)
+				return
+			}
+			// serve
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+func validateRequest(r *http.Request, router routers.Router, options *openapi3filter.Options) (int, error) {
+	// Find route
+	route, pathParams, err := router.FindRoute(r)
+	if err != nil {
+		return http.StatusBadRequest, err // We failed to find a matching route for the request.
```

Comment:
Isn't `NotFound` a better error here?

Reply:
It is based on the middleware implementation, when the request is invalid - doesn't match the swagger schema vs not found which is when the api is valid and the resource is not found (branch, repository, tag and etc...)

Reply:
We use NotFound when a resource in the request is not found like branch, repo or tag and BadRequest (400) when the request is not valid.

Reply:
REST says there is no such difference; I have to agree with @itaiad200 here.  `400` is always appropriate (because it says so little), however accessing a URL that is not mounted is a classic 404.